### DISCUSSION TITLE
How to process *.o3d3xximg file with Python 3?

### DISCUSSION THREAD

**Discussion by peneroka** [ifm/ifm3d#407] [Q&A]
Hi!

I'm looking for ways to process the *.o3d3xximg file (generated by ifm Vision Assistant) using Python 3 (preferably) or any other language available. I want to do offline processing, since I was only given that data and have no access to the camera (type O3D305).

Is there any example/docs/library I could look into? I have tried to look at the [ifm3d Python API](https://api.ifm3d.com/html/_autosummary/ifm3dpy.html#module-ifm3dpy), but I think it was only to be used for online connection to the camera and processing data retrieved from it. I also have looked into [this](https://www.ifm.com/mounting/7391232UK.pdf) and [this](https://www.ifm.com/mounting/706401UK.pdf) doc, and didn't found a clue.

Kindly correct if I'm missing something. Thanks in advance!

**Comment by dekhanra** [comment]
Hi @peneroka,

The ifmVisionAssistant can save the data in HDF5 format and you can unpack the data using h5py (python library), post process the data using ifm3d API.

**Reply by peneroka** [reply]
Thanks for the response, @dekhanra !

I think one can only save to HDF5 when it was first recorded (still connected to the camera)? I didn't found a way to open the existing *.o3d3xximg then perform something like "Save As" to HDF5 format in iVA.

Also from the [docs](https://www.ifm.com/mounting/706401UK.pdf) (pg. 85), it seems that one can only open/load the *.o3d3xximg file during live connection to the camera? The "Replay" menu in iVA only able to open *.h5 and *.dat file.

Is there a way to convert/open/process the already saved *.o3d3xximg file fully offline (without any connection to the camera again)?

**Reply by dekhanra** [reply]
Hi @peneroka,

The .o3d3xximg file is a snapshot of an application and it can be be loaded via iVA while you were connected to the camera only. If there is no access to the camera then you cannot load it and you cannot process it offline. The data is saved in the special format that can be processed by ifmVisionAssistant only.

So, I suggest you to request the data in HDF5 format  from the sender and then you can process it offline using h5py(Python Library). 
To record data in a HDF5 format, Navigate to `Monitor` window on left hand side of ifmVisionAssistant and click `Record options` under the main window. Select the duration and click 
![image](https://github.com/ifm/ifm3d/assets/123174684/14989c3b-4c2c-490b-bf90-449ce7a1bf62)


You can refer to the python code snippet [here](https://ifm3d.com/documentation/ODS/Recording/ods_data_recording.html#how-to-post-process-the-recorded-data) on how to read the HdF5 data file.

**Reply by peneroka** [reply]
Hi again @dekhanra, thanks for the swift response and link to example of HDF5 processing.

I'm still interested with the *.o3d3xximg offline processing, because it is hard to request my data provider to re-record the data as *.h5 with exact replication of the scene/environment identical as before.

> "The .o3d3xximg file is a snapshot of an application"

It also means the captured 3D data is still there, right? Why would we need to have connection to the camera to open it?

> "The data is saved in the special format that can be processed by ifmVisionAssistant only."

I'm curious why ifm does not provide ways to openly process the *.o3d3xximg file? I mean, why differs it with the *.h5 file? Maybe put this in the [ideas](https://github.com/ifm/ifm3d/discussions/categories/ideas) or [polls](https://github.com/ifm/ifm3d/discussions/categories/polls) category?

**Reply by graugans** [reply]
> "The data is saved in the special format that can be processed by ifmVisionAssistant only."

I guess a more correct wording would be a combination of the ifm VisionAssistant and the camera. The idea of this image is more to fine tune some of the application configurations without the need to retrigger the whole scene. This may comes handy in a situation where you have to manually trigger an image in a production line. It was not meant as an exchange format. It contains proprietary ifm internal data which is not useful outside of the camera.

For sharing data and post processing please use the `*.h5` file as recommended by @dekhanra

**Reply by desengph** [reply]
For clarification, one more comment:
Unfortunately, you cannot edit the recorded `o3d3xximg` file. 
The only solution is to re-record the data using the methods described above.

We're sorry for the additional work and possible misunderstandings that occurred before the original data recording.
